[clang][DependencyScanning] Add API to expose module CompilerInvocation

cachemeifyoucan opened the pull request:
Add a new API that returns underlying CompilerInvocation from ModuleDep. This allows clients to inspect and modify the state inside CompilerInvocation before turning that into build arguments.

jansvoboda11 commented on clang/lib/Tooling/DependencyScanning/ModuleDepCollector.cpp:
```diff
@@ -956,9 +956,12 @@ ModuleDepCollectorPP::handleTopLevelModule(const Module *M) {
   // Finish the compiler invocation. Requires dependencies and the context hash.
   MDC.addOutputPaths(CI, MD);
 
+  if (MDC.TransformInvocation)
+    MDC.TransformInvocation(CI);
```
This should definitely run before `associateWithContextHash(CI, ...)`, `areOptionsInStableDir(CI, ...)` and `createCompileJobCacheKey(..., CI)`. What kind of transformation would clients typically perform?

cachemeifyoucan replied:
For the case for swift, it runs after and all cache keys are invalidated. It needs to run in the end, or have more information, like a customized hash function passed (even before scanning).

jansvoboda11 commented on clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:
```diff
@@ -569,7 +569,8 @@ class DependencyScanningAction : public tooling::ToolAction {
           Service, std::move(Opts), ScanInstance, Consumer, Controller,
-          OriginalInvocation, std::move(PrebuiltModuleVFSMap));
+          OriginalInvocation, std::move(PrebuiltModuleVFSMap),
+          Service.getCompilerInvocationTransformation());
```
The entire `Service` is being passed, no need to pass the transformation separately.

jansvoboda11 commented:
Why don't we just expose the underlying `CompilerInvocation` inside `ModuleDeps` to Swift, if it's currently happy modifying it after the Clang scan finishes?

cachemeifyoucan commented:
> Why don't we just expose the underlying CompilerInvocation inside ModuleDeps to Swift, if it's currently happy modifying it after the Clang scan finishes?

Isn't this patch just did something similar? Just need an entry point to modified the underlying `CompilerInvocation` before calling `getBuildArguments()`. Register in the service is less likely to cause error due to calling order of the function of conflicting callback being applied cause the cached build command out of sync. I am definitely up for better alternative if it can make API clearer.

jansvoboda11 commented:
> > Why don't we just expose the underlying CompilerInvocation inside ModuleDeps to Swift, if it's currently happy modifying it after the Clang scan finishes?
> 
> Isn't this patch just did something similar? Just need an entry point to modified the underlying `CompilerInvocation` before calling `getBuildArguments()`. Register in the service is less likely to cause error due to calling order of the function of conflicting callback being applied cause the cached build command out of sync. I am definitely up for better alternative if it can make API clearer.

I find that post-processing value-type results is much clearer compared to callbacks. I also don't think the current implementation of the callback is sound due to the ordering I mentioned in my inlined comment. If that ordering is indeed necessary, I'd prefer to delegate that to the client rather than bake that into the Clang scanner API.

cachemeifyoucan commented:
> If that ordering is indeed necessary, I'd prefer to delegate that to the client rather than bake that into the Clang scanner API.

Ordering is still not controlled by client if you expose internal compiler invocation because you cannot modify the invocation where you mentioned in the inline comment. You can only do it afterwards. But I am fine with adding that flexibility.

cachemeifyoucan committed:
[clang][DependencyScanning] Add API to expose module CompilerInvocation

Add a new API that returns underlying CompilerInvocation from ModuleDep.
This allows clients to inspect and modify the state inside
CompilerInvocation before turning that into build arguments.

(cherry picked from commit 9651a13c528d12d4dfdaae15d751182599f66282)

cachemeifyoucan commented:
@swift-ci please test

cachemeifyoucan commented:
@swift-ci please test

jansvoboda11 reviewed (approved):
LGTM, but this should probably target `swift/release/6.2`, no?

cachemeifyoucan commented:
I have a swift change that I need to land on `main`. I guess I can land this on 6.2 and wait for auto merger to bring it to stable branch before landing swift change. Let me just do it step by step and use another PR to bring to 6.2 branch.

cachemeifyoucan commented:
@swift-ci please test macOS platform

cachemeifyoucan commented:
This is merged in 6.2 already https://github.com/swiftlang/llvm-project/pull/10714

This PR is not needed anymore. Closing.
